core/src/main/java/com/linecorp/armeria/server/auth/HttpAuthServiceBuilder.java

```diff
@@ -57,26 +59,64 @@ public HttpAuthServiceBuilder add(Iterable<? extends Authorizer<HttpRequest>> au
      * Adds an HTTP basic {@link Authorizer}.
      */
     public HttpAuthServiceBuilder addBasicAuth(Authorizer<? super BasicToken> authorizer) {
-        authorizers.add(
-                tokenAuthorizer(AuthTokenExtractors.BASIC, requireNonNull(authorizer, "authorizer")));
-        return this;
+        return addTokenAuthorizer(AuthTokenExtractors.BASIC,
+                                  requireNonNull(authorizer, "authorizer"));
+    }
+
+    /**
+     * Adds an HTTP basic {@link Authorizer} for the given {@code header}.
+     */
+    public HttpAuthServiceBuilder addBasicAuth(Authorizer<? super BasicToken> authorizer, AsciiString header) {
+        return addTokenAuthorizer(new BasicTokenExtractor(requireNonNull(header, "header")),
+                                  requireNonNull(authorizer, "authorizer"));
     }
 
     /**
      * Adds an OAuth1a {@link Authorizer}.
      */
     public HttpAuthServiceBuilder addOAuth1a(Authorizer<? super OAuth1aToken> authorizer) {
-        authorizers.add(
-                tokenAuthorizer(AuthTokenExtractors.OAUTH1A, requireNonNull(authorizer, "authorizer")));
-        return this;
+        return addTokenAuthorizer(AuthTokenExtractors.OAUTH1A,
+                                  requireNonNull(authorizer, "authorizer"));
+    }
+
+    /**
+     * Adds an OAuth1a {@link Authorizer} for the given {@code header}.
+     */
+    public HttpAuthServiceBuilder addOAuth1a(Authorizer<? super OAuth1aToken> authorizer, AsciiString header) {
+        return addTokenAuthorizer(new OAuth1aTokenExtractor(header),
```

Comment:
`requireNonNull(header)`? or adding `requireNonNull` to the constructor of each extractor?

Reply:
Oops somehow missed this one - thanks. I think we usually add `requireNonNull` to public APIs and not so much to internal, especially package-local, ones so will leave them here.